lib/message_bus/backends/redis.rb

```diff
@@ -90,104 +87,35 @@ def expire_all_backlogs!
-      local redis_channel_name = KEYS[5]
-
-      local global_id = redis.call("INCR", global_id_key)
-      local backlog_id = redis.call("INCR", backlog_id_key)
-      local payload = table.concat({ global_id, backlog_id, start_payload }, "|")
-      local global_backlog_message = table.concat({ backlog_id, channel }, "|")
-
-      redis.call("ZADD", backlog_key, backlog_id, payload)
-      redis.call("EXPIRE", backlog_key, max_backlog_age)
-      redis.call("ZADD", global_backlog_key, global_id, global_backlog_message)
-      redis.call("EXPIRE", global_backlog_key, max_backlog_age)
-      redis.call("PUBLISH", redis_channel_name, payload)
-
-      redis.call("EXPIRE", backlog_id_key, max_backlog_age)
-
-      if backlog_id > max_backlog_size and backlog_id % clear_every == 0 then
-        redis.call("ZREMRANGEBYSCORE", backlog_key, 1, backlog_id - max_backlog_size)
-      end
-
-      if global_id > max_global_backlog_size and global_id % clear_every == 0 then
-        redis.call("ZREMRANGEBYSCORE", global_backlog_key, 1, global_id - max_global_backlog_size)
-      end
-
-      return backlog_id
-LUA
-
-      LUA_PUBLISH_SHA1 = Digest::SHA1.hexdigest(LUA_PUBLISH)
-
       # (see Base#publish)
       def publish(channel, data, opts = nil)
-        queue_in_memory = (opts && opts[:queue_in_memory]) != false
-
         max_backlog_age = (opts && opts[:max_backlog_age]) || self.max_backlog_age
         max_backlog_size = (opts && opts[:max_backlog_size]) || self.max_backlog_size
 
         redis = pub_redis
         backlog_id_key = backlog_id_key(channel)
         backlog_key = backlog_key(channel)
 
-        msg = MessageBus::Message.new nil, nil, channel, data
-
-        cached_eval(
-          redis,
-          LUA_PUBLISH,
-          LUA_PUBLISH_SHA1,
-          argv: [
-            msg.encode_without_ids,
-            max_backlog_age,
-            max_backlog_size,
-            max_global_backlog_size,
-            channel,
-            clear_every
-          ],
-          keys: [
-            global_id_key,
-            backlog_id_key,
-            backlog_key,
-            global_backlog_key,
-            redis_channel_name
-          ]
-        )
-      rescue ::Redis::CommandError => e
-        if queue_in_memory && e.message =~ /READONLY/
-          @lock.synchronize do
-            @in_memory_backlog << [channel, data]
-            if @in_memory_backlog.length > @max_in_memory_publish_backlog
-              @in_memory_backlog.delete_at(0)
-              @logger.warn("Dropping old message cause max_in_memory_publish_backlog is full: #{e.message}\n#{e.backtrace.join('\n')}")
-            end
+        backlog_id = nil
+        redis.multi do
+          global_id = redis.incr(global_id_key)
+          backlog_id = redis.incr(backlog_id_key)
+          msg = MessageBus::Message.new global_id, backlog_id, channel, data
+          payload = msg.encode
+          redis.zadd(backlog_key, backlog_id, payload)
```

Comment:
We can count a `backlog_id` there as a stream revision in the EventStore db. We certainly don't want to have multiple records with the same "revision".
Arek, if you take a closer look - you will see that I only re-wrote [lua implementation](https://github.com/discourse/message_bus/blob/main/lib/message_bus/backends/redis.rb#L96-L133). I didn't introduce anything new.

Reply:
As we disucss situation is not existing in this case because every payload is a global_id(uniq) + payload. So event if we publish 2 same messages thanks to the global_id they would be different.